### ISSUE TITLE
`FastThreadLocal` does not have the same behavior with Java's `ThreadLocal`

### ISSUE THREAD

**Issue by laosijikaichele** [netty/netty#11937]
### Expected behavior
`FastThreadLocal` should have the same behavior with Java's `ThreadLocal`.
### Actual behavior
`FastThreadLocal` does not have the same behavior with Java's `ThreadLocal`.
### Minimal yet complete reproducer code (or URL to code)
```
import io.netty.util.concurrent.FastThreadLocal;

    public static void main(String[] args) {
        /**
         * The 'for loop' simulates following scene:
         * The application continuously create 'FastThreadLocal' instances, but GC-ed them subsequently.
         * The application running for a long time.
         */
        for (long i = 0; i < Integer.MAX_VALUE; i++) {
            FastThreadLocal<Object> fst_1 = new FastThreadLocal<>();
            // ... do some business logic with 'fst_1'... //
            fst_1 = null; // GC-ed 'fst_1'.
        }

        // One day, suddenly the application can not create FastThreadLocal instance anymore,
        // unless restart the JVM.
        FastThreadLocal<Object> fst_2 = new FastThreadLocal<>(); // Throws Exception
    }
```
The code above simulates a specific scene with `FastThreadLocal`,  it throws exception after run.
If we replace `FastThreadLocal` with Java 's `ThreadLocal`, re-run the code above, it works fine with no error.
### Netty version
4.1.72.Final
### JVM version (e.g. `java -version`)
8
### OS version (e.g. `uname -a`)
MacOS

**Comment by normanmaurer** [comment]
@laosijikaichele We could also "fallback" to `ThreadLocal` once we reach the max index.

**Comment by amizurov** [comment]
I'm wondering why we need to create so many `FastThreadLocal` threads, for me it's look like we use it an incorrect way. And if we switch to "fallback" it will be implicit behaviour and may lead to another problems, IMHO.

**Comment by normanmaurer** [comment]
Yeah it definitely looks like an anti-pattern

**Comment by franz1981** [comment]
> I'm wondering why we need to create so many FastThreadLocal threads, for me it's look like we use it an incorrect way. And if we switch to "fallback" it will be implicit behaviour and may lead to another problems, IMHO.

I saw things like these to appear while using `FastThreadLocal` on unbounded thread pools that can remove threads while idle or can let the thread pool to grow to accomodate some burst of load with the existing threads alive blocked doing "something"

**Comment by laosijikaichele** [comment]
> @laosijikaichele We could also "fallback" to `ThreadLocal` once we reach the max index.

@normanmaurer Yep, seems to be the solution 3?

**Comment by laosijikaichele** [comment]
> Yeah it definitely looks like an anti-pattern

@normanmaurer , @amizurov , I agree it is an ab-normal user case, but I consider this issue more on framework/application robust perspective, and also the compatibility with Java's `ThreadLocal`.

**Comment by normanmaurer** [comment]
@laosijikaichele yes
